File: services/notification-service/lib/notification_service.rb
require_relative 'database'
require_relative '../protos/notification_services_pb'

class NotificationServiceImpl < Notification::NotificationService::Service
  def send_order_notification(request, _call)
    puts "[NotificationService] Sending order notification for order #{request.order_id}"

    message = "Order #{request.order_id} for #{request.customer_name} - Status: #{request.status}, Total: $#{request.total_amount}"

    notification_id = save_notification('ORDER', message)

    puts "[NotificationService] Notification saved with ID: #{notification_id}"

    Common::StatusResponse.new(
      success: true,
      message: "Order notification sent successfully"
    )
  rescue => e
    puts "[NotificationService] Error: #{e.message}"
    Common::StatusResponse.new(
      success: false,
      message: "Failed to send notification: #{e.message}"
    )
  end

  def send_stock_alert(request, _call)
    puts "[NotificationService] Sending stock alert for product #{request.product_id}"

    message = "Stock Alert [#{request.alert_type}]: #{request.product_name} - Current quantity: #{request.current_quantity}"

    notification_id = save_notification('STOCK_ALERT', message)

    puts "[NotificationService] Stock alert saved with ID: #{notification_id}"

    Common::StatusResponse.new(
      success: true,
      message: "Stock alert sent successfully"
    )
  rescue => e
    puts "[NotificationService] Error: #{e.message}"
    Common::StatusResponse.new(
      success: false,
      message: "Failed to send stock alert: #{e.message}"
    )
  end

  def get_notifications(request, _call)
    puts "[NotificationService] Fetching notifications"

    limit = request.limit > 0 ? request.limit : 10
    offset = request.offset > 0 ? request.offset : 0

    total = Database.notifications.count

    notifications = Database.notifications
      .order(Sequel.desc(:created_at))
      .limit(limit)
      .offset(offset)
      .all
      .map do |row|
        Notification::Notification.new(
          id: row[:id].to_s,
          type: row[:type],
          message: row[:message],
          created_at: row[:created_at].iso8601,
          sent: row[:sent]
        )
      end

    puts "[NotificationService] Found #{notifications.length} notifications"

    Notification::NotificationsResponse.new(
      notifications: notifications,
      total: total
    )
  rescue => e
    puts "[NotificationService] Error: #{e.message}"
    Notification::NotificationsResponse.new(
      notifications: [],
      total: 0
    )
  end

  private

  def save_notification(type, message)
    id = SecureRandom.uuid
    Database.notifications.insert(
      id: id,
      type: type,
      message: message,
      sent: true,
      created_at: Time.now
    )
    id
  end
end

File: services/notification-service/internal/service/notification_service.go
package service

import (
	"context"
	"fmt"
	"log"

	"notification-service/internal/database"
	pb "notification-service/proto"
)

type NotificationServiceServer struct {
	pb.UnimplementedNotificationServiceServer
	db *database.DB
}

func NewNotificationServiceServer(db *database.DB) *NotificationServiceServer {
	return &NotificationServiceServer{db: db}
}

func (s *NotificationServiceServer) SendOrderNotification(ctx context.Context, req *pb.OrderNotificationRequest) (*pb.StatusResponse, error) {
	message := fmt.Sprintf("Order %s for %s - Status: %s, Total: $%.2f",
		req.OrderId, req.CustomerName, req.Status, req.TotalAmount)

	id, err := s.db.SaveNotification("ORDER", message)
	if err != nil {
		log.Printf("Failed to save order notification: %v", err)
		return &pb.StatusResponse{
			Success: false,
			Message: fmt.Sprintf("Failed to send notification: %v", err),
		}, nil
	}

	log.Printf("Order notification sent: %s", id)
	return &pb.StatusResponse{
		Success: true,
		Message: fmt.Sprintf("Notification sent successfully with ID: %s", id),
	}, nil
}

func (s *NotificationServiceServer) SendStockAlert(ctx context.Context, req *pb.StockAlertRequest) (*pb.StatusResponse, error) {
	message := fmt.Sprintf("Stock Alert [%s]: %s - Current quantity: %d",
		req.AlertType, req.ProductName, req.CurrentQuantity)

	id, err := s.db.SaveNotification("STOCK_ALERT", message)
	if err != nil {
		log.Printf("Failed to save stock alert: %v", err)
		return &pb.StatusResponse{
			Success: false,
			Message: fmt.Sprintf("Failed to send alert: %v", err),
		}, nil
	}

	log.Printf("Stock alert sent: %s", id)
	return &pb.StatusResponse{
		Success: true,
		Message: fmt.Sprintf("Alert sent successfully with ID: %s", id),
	}, nil
}

func (s *NotificationServiceServer) GetNotifications(ctx context.Context, req *pb.GetNotificationsRequest) (*pb.NotificationsResponse, error) {
	notifications, total, err := s.db.GetNotifications(req.Limit, req.Offset)
	if err != nil {
		log.Printf("Failed to get notifications: %v", err)
		return &pb.NotificationsResponse{
			Notifications: []*pb.Notification{},
			Total:         0,
		}, nil
	}

	pbNotifications := make([]*pb.Notification, len(notifications))
	for i, n := range notifications {
		pbNotifications[i] = &pb.Notification{
			Id:        n.ID,
			Type:      n.Type,
			Message:   n.Message,
			CreatedAt: n.CreatedAt.Format("2006-01-02T15:04:05Z"),
			Sent:      n.Sent,
		}
	}

	return &pb.NotificationsResponse{
		Notifications: pbNotifications,
		Total:         total,
	}, nil
}
